File: NeuralNetwork/CatsVsDogs.cs
using System.Drawing;
using static Neural_Network_and_AI.MathUtil;
namespace Neural_Network_and_AI
{
    public class CatsVsDogs : Dataset
    {
        private readonly string[] categories =
        [
            "Cat",
            "Dog"
        ];
        private int compressedWidth = 32;
        private int compressedHeight = 32;
        public void Testing(NeuralNetwork nn)
        {
            Console.WriteLine("Input path to a photo of a cat or dog:");
            var inputPath = Console.ReadLine();
            Bitmap bitmap = new Bitmap(inputPath);
            var floatArray = ImageProcessing.GetPixelColorsAsFloatArray(bitmap, compressedWidth, compressedHeight);
            var results = nn.Predict(floatArray.ToList(), NeuralNetwork.PredictionMethod.Softmax);
            Console.WriteLine($"Probability of photo being a Cat: {results[0] * 100}%");
            Console.WriteLine($"Probability of photo being a Dog: {results[1] * 100}%");
        }

        public (List<List<float>>, List<List<float>>) TrainingData(int samples = -1)
        {
            List<Tuple<string, List<float>>> imageFloatArrays;
            Dictionary<string, float[]> categoryToOneHot = categories.ToDictionary(cat => cat, cat => GetOneHotEncodedArr(cat, categories));

            imageFloatArrays = Dataset.LoadDataFromFileAll("PetImages", ".jpg", ImageProcessing.GetPixelColorsAsFloatArray, compressedWidth, compressedHeight, categories, samples);

            Random random = new Random(15235);
            List<List<float>> allInputs = new List<List<float>>();
            List<List<float>> allExpected = new List<List<float>>();

            //Shuffle the images so that we get random and not all of one category
            var shuffled = imageFloatArrays.OrderBy(x => random.Next()).ToList();

            foreach (var (category, imageArray) in shuffled.Take(samples))
            {
                allInputs.Add(imageArray);
                allExpected.Add(categoryToOneHot[category].ToList());
            }

            Console.WriteLine($"Total images loaded: {imageFloatArrays.Count}, Total sampled: {samples}");

            return (allInputs, allExpected);
        }
    }
}

File: NeuralNetwork/Python/CatsVsDogs.py
from Dataset import Dataset
import random
from PIL import Image
from ImageProcessing import ImageProcessing
from NeuralNetwork import NeuralNetwork
from MathUtil import MathUtil
class CatsVsDogs(Dataset):

    def __init__(self):
        self.categories = [
            "Cat",
            "Dog"
        ]

        self.compressedWidth = 32
        self.compressedHeight = 32

    def Testing(self, nn):

        print("Input path to a photo of a cat or dog:")
        inputPath = input()

        bitmap = Image.open(inputPath)

        floatArray = ImageProcessing.GetPixelColorsAsFloatArray(
            bitmap,
            self.compressedWidth,
            self.compressedHeight
        )

        results = nn.Predict(
            floatArray,
            NeuralNetwork.PredictionMethod.Softmax
        )

        print(f"Probability of photo being a Cat: {results[0] * 100}%")
        print(f"Probability of photo being a Dog: {results[1] * 100}%")

    def TrainingData(self, samples=-1):

        categoryToOneHot = {
            cat: MathUtil.GetOneHotEncodedArr(cat, self.categories)
            for cat in self.categories
        }

        imageFloatArrays = Dataset.LoadDataFromFileAll(
            "../PetImages",
            ".jpg",
            self.compressedWidth,
            self.compressedHeight,
            self.categories,
            samples
        )

        allInputs = []
        allExpected = []

        random.seed(15235)
        random.shuffle(imageFloatArrays)

        # If samples == -1, take all
        if samples == -1:
            selected = imageFloatArrays
        else:
            selected = imageFloatArrays[:samples]

        for category, imageArray in selected:
            allInputs.append(imageArray)
            allExpected.append(list(categoryToOneHot[category]))

        print(f"Total images loaded: {len(imageFloatArrays)}, Total sampled: {samples}")

        return (allInputs, allExpected)
